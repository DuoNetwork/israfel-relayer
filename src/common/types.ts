import { SignedOrder } from '0x.js';

export interface ILiveOrder {
	account: string;
	pair: string;
	orderHash: string;
	price: number;
	amount: number;
	balance: number;
	fill: number;
	side: string;
	createdAt?: number;
	updatedAt?: number;
	initialSequence: number;
	currentSequence: number;
}

export interface IUserOrder extends ILiveOrder {
	type: string;
	status: string;
	updatedBy: string;
}

export interface IRawOrder {
	orderHash: string;
	signedOrder: IStringSignedOrder | SignedOrder;
	createdAt?: number;
	updatedAt?: number;
}

export interface IStringSignedOrder {
	senderAddress: string;
	makerAddress: string;
	takerAddress: string;
	makerFee: string;
	takerFee: string;
	makerAssetAmount: string;
	takerAssetAmount: string;
	makerAssetData: string;
	takerAssetData: string;
	salt: string;
	exchangeAddress: string;
	feeRecipientAddress: string;
	expirationTimeSeconds: string;
	signature: string;
}

export interface IOrderPersistRequest {
	method: string;
	pair: string;
	orderHash: string;
	balance: number;
	signedOrder?: IStringSignedOrder;
}

export interface IOrderQueueItem {
	method: string;
	liveOrder: ILiveOrder;
	signedOrder?: IStringSignedOrder;
}

export interface IOrderBookSnapshotWs extends IOrderBookSnapshot {
	type: string;
	sequence: number;
	channel: string;
}

export interface IOrderBookSnapshot {
	sequence: number;
	bids: IOrderBookUpdateWS[];
	asks: IOrderBookUpdateWS[];
}

export interface IUpdatePayloadWs {
	order: IStringSignedOrder;
	metaData: {
		remainingTakerAssetAmount: string;
	};
}

export interface IOrderBookUpdateWS {
	price: number;
	amount: number;
}

export interface IOrderBookUpdate {
	sequence: number;
	pair: string;
	price: number;
	amount: number;
}

export interface IWsRequest {
	method: string;
	channel: string;
	pair: string;
}

export interface IWsResponse {
	status: string;
	method: string;
	channel: string;
	pair: string;
}

export interface IWsOrderRequest extends IWsRequest {
	orderHash: string;
}

export interface IWsAddOrderRequest extends IWsOrderRequest {
	order: IStringSignedOrder | SignedOrder;
}

export interface IWsOrderResponse extends IWsResponse {
	orderHash: string;
}

export interface IWsUserOrderResponse extends IWsOrderResponse {
	userOrder: IUserOrder;
}

export interface ISubscribeOrderBookRequest {
	method: string;
	channel: string;
}

// export interface IUpdateResponseWs {
// 	type: WsChannelResposnseTypes;
// 	lastTimestamp: number;
// 	currentTimestamp: number;
// 	channel: {
// 		name: WsChannelName;
// 		pair: string;
// 	};
// 	bids: IOrderBookUpdateWS[];
// 	asks: IOrderBookUpdateWS[];
// }

// export interface IOrderResponseWs {
// 	channel: {
// 		name: WsChannelName;
// 		pair: string;
// 	};
// 	status: string;
// 	failedReason: string;
// }

export interface IOption {
	live: boolean;
	token: string;
	maker: number;
	spender: number;
	amount: number;
	debug: boolean;
	server: boolean;
}

export interface IStatus {
	hostname: string;
	updatedAt: number;
	pair: string;
	tool: string;
	count?: number;
}

export interface IService {
	service: string;
	hostname: string;
	url: string;
}
