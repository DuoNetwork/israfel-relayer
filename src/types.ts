import { BigNumber, ExchangeContractErrs, OrderRelevantState, SignedOrder } from '0x.js';

export interface IDuoOrder extends IDuoSignedOrder {
	orderHash: string;
	isValid: boolean;
	isCancelled: boolean;
	updatedAt: number;
	orderWatcherState: OrderRelevantState | ExchangeContractErrs;
}

// export interface IOrderBook {
// 	bids: IDuoOrder[];
// 	asks: IDuoOrder[];
// }

export interface IOrderBookSnapshotWs {
	type: WsChannelResposnseTypes;
	timestamp: number;
	channel: {
		name: string;
		marketId: string;
	};
	requestId: number;
	bids: IOrderBookUpdateWS[];
	asks: IOrderBookUpdateWS[];
}

export interface IUpdatePayloadWs {
	order: SignedOrder;
	metaData: {
		remainingTakerAssetAmount: BigNumber;
	};
}

export interface IOrderBookUpdateWS {
	price: string;
	amount: string;
}

export interface IUpdateResponseWs {
	type: WsChannelResposnseTypes;
	lastTimestamp: number;
	currentTimestamp: number;
	channel: {
		name: WsChannelName;
		marketId: string;
	};
	bids: IOrderBookUpdateWS[];
	asks: IOrderBookUpdateWS[];
}

export interface IOrderResponseWs {
	channel: {
		name: WsChannelName;
		marketId: string;
	};
	status: string;
	failedReason: string;
}

export interface ICancelOrderResponseWs {
	status: string;
	orderHash: string;
}

export enum ErrorResponseWs {
	InvalidOrder = 'Invalid order schema or signature!',
	NoExistOrder = 'Order does not exist in DB!'
}

export interface IDuoSignedOrder {
	senderAddress: string;
	makerAddress: string;
	takerAddress: string;
	makerFee: string;
	takerFee: string;
	makerAssetAmount: string;
	takerAssetAmount: string;
	makerAssetData: string;
	takerAssetData: string;
	salt: string;
	exchangeAddress: string;
	feeRecipientAddress: string;
	expirationTimeSeconds: string;
	signature: string;
}

export enum WsChannelMessageTypes {
	Add = 'add',
	Cancel = 'cancel',
	Subscribe = 'subscribe'
}

export enum WsChannelResposnseTypes {
	Update = 'update',
	Snapshot = 'snapshot'
}

export enum WsChannelName {
	Orderbook = 'orderbook',
	Order = 'order'
}

export interface IOrderStateCancelled {
	isCancelled: boolean;
	orderHash: string;
}

export interface IOption {
	token: string;
	maker: number;
	spender: number;
	amount: number;
}
