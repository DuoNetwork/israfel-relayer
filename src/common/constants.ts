export const DB_ISRAFEL = 'israfel';
export const DB_LIVE_ORDERS = 'liveOrders';
export const DB_RAW_ORDERS = 'rawOrders';
export const DB_USER_ORDERS = 'userOrders';
export const DB_SEQUENCE = 'sequence';
export const DB_STATUS = 'status';
export const DB_LIVE = 'live';
export const DB_DEV = 'dev';
export const DB_PAIR = 'pair';
export const DB_INITIAL_SEQ = 'initialSequence';
export const DB_CURRENT_SEQ = 'currentSequence';
export const DB_ORDER_HASH = 'orderHash';
export const DB_PRICE = 'price';
export const DB_ACCOUNT = 'account';
export const DB_ACCOUNT_YM = 'accountYearMonth';
export const DB_PAIR_OH_SEQ_STATUS = 'pairOrderHashSequenceStatus';
export const DB_TYPE = 'type';
export const DB_CONFIRMED = 'confirmed';
export const DB_MATCHING = 'matching';
export const DB_USER = 'user';
export const DB_ORDER_PROCESSOR = 'orderProcessor';
export const DB_ORDER_MATCHER = 'orderMatcher';
export const DB_ORDER_WATCHER = 'orderWatcher';
export const DB_RELAYER = 'relayer';
export const DB_URL = 'url';
export const DB_SIDE = 'side';
export const DB_ASK = 'ask';
export const DB_BID = 'bid';
export const DB_BALANCE = 'balance';
export const DB_AMOUNT = 'amount';
export const DB_FILL = 'fill';
export const DB_PFILL = 'pFill';
export const DB_CREATED_AT = 'createdAt';
export const DB_UPDATED_AT = 'updatedAt';
export const DB_UPDATED_BY = 'updatedBy';
export const DB_PROCESSED = 'processed';
export const DB_TOKENS = 'tokens';
export const DB_ADDRESS = 'address';
export const DB_CODE = 'code';
export const DB_EXP = 'expiry';
export const DB_DENOMINATION = 'denomination';
export const DB_PRECISION = 'precision';
export const DB_FEE = 'fee';
export const DB_FEE_ASSET = 'feeAsset';

export const DB_ADD = 'add';
export const DB_UPDATE = 'update';
export const DB_TERMINATE = 'terminate';

export const DB_PROCESS = 'process';
export const DB_HOSTNAME = 'hostname';
export const DB_COUNT = 'count';

export const DB_0X_SENDER_ADDR = 'senderAddress';
export const DB_0X_MAKER_FEE = 'makerFee';
export const DB_0X_TAKER_FEE = 'takerFee';
export const DB_0X_MAKER_ADDR = 'makerAddress';
export const DB_0X_TAKER_ADDR = 'takerAddress';
export const DB_0X_MAKER_ASSET_AMT = 'makerAssetAmount';
export const DB_0X_TAKER_ASSET_AMT = 'takerAssetAmount';
export const DB_0X_MAKER_ASSET_DATA = 'makerAssetData';
export const DB_0X_TAKER_ASSET_DATA = 'takerAssetData';
export const DB_0X_SALT = 'salt';
export const DB_0X_EXCHANGE_ADDR = 'exchangeAddress';
export const DB_0X_FEE_RECIPIENT_ADDR = 'feeRecipientAddress';
export const DB_0X_EXPIRATION_TIME_SECONDS = 'expirationTimeSeconds';
export const DB_0X_SIGNATURE = 'signature';

export const DB_ORDERS = 'orders';
export const DB_ORDER_BOOKS = 'orderBooks';
export const DB_QUEUE = 'queue';
export const DB_CACHE = 'cache';
export const DB_PUBSUB = 'pubsub';
export const DB_SNAPSHOT = 'snapshot';

export const AWS_DYNAMO_API_VERSION = '2012-10-08';

export const TAKER_ETH_DEPOSIT = 10; // for development only
export const ONE_SECOND_MS = 1000;
export const ONE_MINUTE_MS = ONE_SECOND_MS * 60;
export const TEN_MINUTES_MS = ONE_MINUTE_MS * 10;
export const RELAYER_ADDR_KOVAN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const RELAYER_ADDR_MAIN = '0x003519A4aB2C35c59Cb31d9194A45DD3F9Bf9e32';
export const DUMMY_ADDR = '0x0000000000000000000000000000000000000000';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const PROVIDER_INFURA_MAIN = 'https://mainnet.infura.io';
export const NETWORK_ID_MAIN = 1;
export const NETWORK_ID_KOVAN = 42;
export const PROVIDER_LOCAL = 'http://localhost:8545';
export const NETWORK_ID_LOCAL = 50;
export const PENDING_HOURS = 24;

export const WS_HISTORY = 'history';
export const WS_INFO = 'info';
export const WS_SUB = 'subscribe';
export const WS_UNSUB = 'unsubscribe';
export const WS_CHANNEL = 'channel';
export const WS_METHOD = 'method';
export const WS_INVALID_REQ = 'invalid request';
export const WS_INVALID_ORDER = 'invalid order';
export const WS_ERROR = 'error';
export const WS_OK = 'ok';
export const WS_SERVICE_NA = 'service not available';

export const WS_CHANNEL_ORDERBOOK = 'orderbook';
export const WS_TYPE_ORDERBOOK = 'subscribe';
export const WS_CHANNEL_ORDER = 'order';
export const WS_TYPE_ORDER_UPDATE = 'update';
export const WS_TYPE_ORDER_ADD = 'add';
export const WS_TYPE_ORDER_CANCEL = 'cancel';

export const TOKEN_WETH = 'WETH';

export const PRUNE_INTERVAL = 3600000;
export const ORDER_PRUNE = 'pruneOrder';
export const SET_ALLOWANCE = 'setAllowance';
export const START_RELAYER = 'startRelayer';

export const LOG_INFO = 'INFO';
export const LOG_DEBUG = 'DEBUG';
export const LOG_ERROR = 'ERROR';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const PRICE_PRECISION = 8;

export const BASE_DERIVATION_PATH = `44'/60'/0'/0`;
