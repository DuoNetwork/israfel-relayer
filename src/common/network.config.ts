
export const mnemonicConfig = {
	MNEMONIC: 'alice bob charles david eric frank george henry is jack kobe let',
	BASE_DERIVATION_PATH: `44'/60'/0'/0`
}

export const KOVAN_CONFIGS = {
	rpcUrl: 'https://kovan.infura.io/WSDscoNUvMiL1M7TvMNP',
	networkId: 42,
};
