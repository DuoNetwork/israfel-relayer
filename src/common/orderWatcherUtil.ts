import { ContractWrappers, OrderWatcher, RPCSubprovider } from '0x.js';
import * as CST from '../constants';
import dynamoUtil from '../dynamoUtil';
import { providerEngine } from '../providerEngine';
import { ILiveOrders, IOption } from '../types';
import util from '../util';

class OrderWatcherUtil {
	public provider = new RPCSubprovider(CST.PROVIDER_LOCAL);
	// public providerEngine = new Web3ProviderEngine();
	public zeroEx: ContractWrappers;
	public orderWatcher: OrderWatcher;

	constructor() {
		// this.providerEngine.addProvider(this.provider);
		// this.providerEngine.start();
		this.zeroEx = new ContractWrappers(providerEngine, { networkId: CST.NETWORK_ID_LOCAL });
		this.orderWatcher = new OrderWatcher(providerEngine, CST.NETWORK_ID_LOCAL);
	}

	public init(tool: string, option: IOption) {
		const config = require('./keys/' + (option.live ? 'live' : 'dev') + '/dynamo.json');
		dynamoUtil.init(config, option.live, tool);
	}

	public unsubOrderWatcher() {
		this.orderWatcher.unsubscribe();
	}

	//remove orders remaining invalid for 24 hours from DB
	public async pruneOrders(option: IOption) {
		const marketId = option.token + '-' + CST.TOKEN_WETH;
		const orders = await dynamoUtil.getLiveOrders(marketId);
		console.log('length before prune is', orders.length);
		orders.forEach(order => {
			const inValidTime = !order.isValid ? Date.now() - order.updatedAt : 0;
			console.log(inValidTime);
			if (inValidTime > CST.PENDING_HOURS * 3600000) {
				dynamoUtil.removeLiveOrder(marketId, order.orderHash);
				this.orderWatcher.removeOrder(order.orderHash);
				console.log('remove order!');
			}
		});
		console.log('length after prune is', orders.length);
	}

	public async startOrderWatcher(option: IOption) {
		const marketId = option.token + '-' + CST.TOKEN_WETH;
		util.logInfo('start order watcher for ' + marketId);
		const liveOrders: ILiveOrders[] = await dynamoUtil.getLiveOrders(marketId);

		console.log('length in DB is ', liveOrders.length);
		for (const order of liveOrders)
			try {
				await this.orderWatcher.addOrderAsync(
					await dynamoUtil.getRawOrder(order.orderHash)
				);
				console.log('succsfully added %s', order.orderHash);
			} catch (e) {
				console.log('failed to add %s', order.orderHash, 'error is ' + e);
			}

		this.orderWatcher.subscribe(async (err, orderState) => {
			if (err) {
				console.log(err);
				return;
			}

			console.log(Date.now().toString(), orderState);
			if (orderState !== undefined) await dynamoUtil.updateOrderState(orderState, marketId);
		});
	}
}

const orderWatcherUtil = new OrderWatcherUtil();
export default orderWatcherUtil;
