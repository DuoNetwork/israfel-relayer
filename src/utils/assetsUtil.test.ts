import {BigNumber} from '0x.js';
import { Web3Wrapper } from '@0xproject/web3-wrapper';
import assetsUtil from './assetsUtil';
import { IOption } from '../common/types';

test('getTokenAddressFromName', async () => {
	expect(assetsUtil.getTokenAddressFromName('ZRX')).toMatchSnapshot();
	expect(assetsUtil.getTokenAddressFromName('WETH')).toMatchSnapshot();
	expect(assetsUtil.getTokenAddressFromName('')).toMatchSnapshot();
});

test('setTokenAllowance', async () => {
	assetsUtil.web3Wrapper.getAvailableAddressesAsync = jest.fn(() =>
		Promise.resolve(['account1', 'account2', 'account3', 'account4'])
	);

	await assetsUtil.init();
	const option: IOption = {
		live: false,
		token: 'ZRX',
		maker: 0,
		spender: 1,
		amount: 0.1,
		debug: false
	};

	Web3Wrapper.toBaseUnitAmount = jest.fn((input: BigNumber, decimal: number) => input.toPower(decimal));
	assetsUtil.contractWrappers.erc20Token.setAllowanceAsync = jest.fn(() =>
		Promise.resolve('orderHash')
	);

	// expect(assetsUtil.setTokenAllowance(option) as jest.Mock<Promise<void>>)
});
