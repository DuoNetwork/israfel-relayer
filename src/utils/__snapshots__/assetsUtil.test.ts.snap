// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTokenAddressFromName 1`] = `"0x871dd7c2b4b25e1aa18728e9d5f2af4c4e431f5c"`;

exports[`getTokenAddressFromName 2`] = `"0x0b1ba0af832d7c05fd64161e0db78e85978e8082"`;

exports[`getTokenAddressFromName 3`] = `""`;
